### ISSUE TITLE
As X-Road Product Owner I want that the global configuration schema can be changed so that the change does not break the backwards compatibility

### ISSUE THREAD

**Issue by hanhaka** [vrk-kpa/xroad-joint-development#69]
**Affected components**: Global configuration
**Affected documentation**: -
**Estimated delivery**: Autumn 2016
**External reference**: https://jira.csc.fi/browse/PVAYLADEV-474

**Problem**
Currently it is not possible to change the global configuration schema without breaking the backwards compatibility. By adding to or removing something from the schema, Central Server and Security Servers cannot communicate together if the Global Conf XML schema is not compatible. This will be causing a big break in X-Road communication when upgrading the new X-Road sw version.

There will be new features in global configuration now and then and that's it is needed to a general way to manage this kind of situation.

**Acceptance criteria**
- Global configuration schema can be changed without breaking the backwards compatibility

**Comment by hanhaka** [comment]
Agreed on 18th of August that work on this issue can be started (done by VRK side).

**Comment by PriitParmakson** [comment]
What do we mean by "backwards compatibility"? It has to be defined or explained. Currently the only extension mechanism provided by "PR-GCONF X-Road: Protocol for Downloading Configuration" (v2.1) is that "X-Road implementations are free to define additional types of configuration parts"  (section 2.5) (in addtion to PRIVATE-PARAMETERS and SHARED-PARAMETERS XML-files). That means that global configuration can contain other files than PRIVATE-PARAMETERS and SHARED-PARAMETERS files. These two files, however, have inflexible schemas (as far as I can decide).  PRIVATE-PARAMETERS and SHARED-PARAMETERS files contain no XML "any" elements. The files also do not contain version numbers.

We should define the backwards compatibility use case(s) first. 

"Global configuration provider introduces a new version of global configuration (say 2.1). The new version is backwards compatible in the sense that user can process it by skipping new elements; elements present in version 2.0 are preserved, including the semantics."

Could this be the use case?

Let me note that we need to develop a similar mechanism for Estonian document exchange XML capsule format.

**Comment by PriitParmakson** [comment]
Designing Schemas for Backward and Forward Compatibility http://www.xfront.com/backward-forward-compatibility/ looks important. If you know other sources for the theme, please drop here.

**Comment by petkivim** [comment]
One alternative would be to define configuration parameters as key value pairs, e.g.:

```
<m:paramSet>
    <m:stringparam>
        <m:name>param1</m:name>
        <m:value>valueq</m:value>
    </m:stringparam>
    <m:paramSet>
        <m:name>systemparams</m:name>
        <m:stringparam>
            <m:name>param2</m:name>
            <m:value>value2</m:value>
        </m:stringparam>
        <m:numericparam>
            <m:name>param3</m:name>
            <m:value>10</m:value>
        </m:numericparam>
        <m:boolparam>
            <m:name>param4</m:name>
            <m:value>true</m:value>
        </m:boolparam>
    </m:paramSet>     
</m:paramSet>
```

In this way new parameters can be added without changing the schema.

**Comment by hanhaka** [comment]
@PriitParmakson This issue was originally raised due to findings in 6.8.4 pull request that Global Configuration was changed in a way that v. 6.8.4 is not anymore compatible with previous sw versions (in v.6.8.4 the ‘identifierDecoder’ –element has been replaced with ‘certificateProfileInfo’ –element and the type of element has also changed => these two changes are breaking the schema). 

We need some general way to do the schema changes now and in the future, and at the same time preserve the compatibility. For example that link provided by you, Designing Schemas for Backward and Forward Compatibility, is a very good starting point for designing the valid schema for X-road.

The ultimate goal of this issue #69 is to find a good solution and document it properly (not only internally but to public documentation => 'pr-gconf_x-road_protocol_for_downloading_configuration_x.doc'). Our developers will look at different solutions and hopefully they will find the best choice for our purposes. The one proposed by Petteri (above) is also one alternative how it could be done.

And yes, I think that your use case definition above for backward compatibility is the one we need:
_"Global configuration provider introduces a new version of global configuration (say 2.1). The new version is backwards compatible in the sense that user can process it by skipping new elements; elements present in version 2.0 are preserved, including the semantics."_

**Comment by PriitParmakson** [comment]
Also of interest may be:
- the [critique](http://vasters.com/clemensv/2014/06/02/MQTT+An+Implementers+Perspective.aspx) by Clemens Vasters of the MQTT protocol (sections Protocol Versioning and Extensibility in particular).
- [Postel's robustness principle](https://en.wikipedia.org/wiki/Robustness_principle) (including the opposite views)
- [RFC 6709 Design Considerations for Protocol Extensions](http://www.ietf.org/rfc/rfc6709.txt) (2012).

**Comment by PriitParmakson** [comment]
Ilkka a.o,
Re: options A, B, and C (in the proposal):
- Would XML/XPath attacks be realistic? Global configuration files are prepared by X-Road Centre - a trusted party? The files are also signed. The files are not unvalidated input from unknown user.
- Would the use of _any_ schema element pose inherent risk? 
- I support strict validation of the schema ("partial" validation would make little sense, I believe).
  (I'm no XML expert. However, similar issues exist in our Document Exchange Capsule schema.)

**Comment by hvainsalu** [comment]
It is WIP?

**Comment by hanhaka** [comment]
Yes it is. We started the implementation during the current Sprint. (I just updated the WIP label to this issue.)

Implementation should be ready after the next Sprint (after two weeks).

**Comment by hanhaka** [comment]
New estimate for the code changes to be ready: after two next Sprints (after four weeks). Still some changes to be made to Configuration Proxy and Security Server side to support several versions of global configuration scheme.

**Comment by hanhaka** [comment]
Commited to XM/develop (finnish-6.9.0 pull request). Will be available with 6.9 release.
